webapp/src/chropro/Renderer.java

```diff
@@ -11,30 +13,37 @@
 
 public class Renderer implements AutoCloseable {
 
-
-    private final Chropro chrome;
+    private Chropro chrome;
     private final String chromeHost;
     private final int chromePort;
+    private String contextId;
+    private int timeout = 10000;
 
-    public Renderer(String chromeHost, int chromePort) throws IOException {
+    public Renderer(String chromeHost, int chromePort) throws IOException, TimeoutException, InterruptedException {
         this.chromeHost = chromeHost;
         this.chromePort = chromePort;
-        this.chrome = new Chropro("ws://" + this.chromeHost + ":" + this.chromePort + "/devtools/browser");
-
+        init();
     }
 
-    public byte[] render(String url, int w, int h, int timeout, int sleep) throws ExecutionException, InterruptedException, IOException, TimeoutException {
-        String contextId;
-
+    private void init() throws IOException, TimeoutException, InterruptedException {
+        chrome = this.chrome = new Chropro("ws://" + this.chromeHost + ":" + this.chromePort + "/devtools/browser");
         try {
             contextId = chrome.target.createBrowserContext().get(timeout, MILLISECONDS).browserContextId;
         } catch (ExecutionException e) {
             contextId = null; // browser contexts are only supported on headless chrome
         }
+    }
 
+    public byte[] render(String url, int w, int h, int timeout, int sleep) throws ExecutionException, InterruptedException, IOException, TimeoutException {
         String targetId = null;
         try {
-            targetId = chrome.target.createTarget("about:blank", w, h, contextId).get(timeout, MILLISECONDS).targetId;
+            try {
+                targetId = chrome.target.createTarget("about:blank", w, h, contextId).get(timeout, MILLISECONDS).targetId;
+            } catch (WebsocketNotConnectedException ex) {
+                System.out.println("Lost connection to web service. Re-initialising chrome client.");
+                init();
```

Comment:
I think there's race here if two requests come in while the connection is dead and both try to re-establish it. The first across the line will have their connection replaced and leak. Maybe we could mark `init()` as 'synchronized' and also call close on the connection that we're replacing?

That can still race but at least it'll just cause a temporary error rather than a connection leak.